### PR TITLE
Update the Rewrite to parse URLs into query vars.

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/128673

This PR also:
* adds the `\Tribe` namespace to the Composer autoload
* adds the `Cache_User` trait
* adds an expiration trigger to the `Tribe__Cache` on `generate_rewrite_rules`

### CODE REVIEW

File: src/Tribe/Cache_Listener.php [hunk 1]
```diff
@@ -4,6 +4,9 @@
 	 */
 	class Tribe__Cache_Listener {
 
+		const TRIGGER_GENERATE_REWRITE_RULES = 'generate_rewrite_rules';
+		const TRIGGER_SAVE_POST = 'save_post';
```
Comment (bordoni) [inline src/Tribe/Cache_Listener.php]: Docblock.

File: src/Tribe/Cache_Listener.php [hunk 2]
```diff
@@ -33,6 +36,7 @@ public function init() {
 		private function add_hooks() {
 			add_action( 'save_post', array( $this, 'save_post' ), 0, 2 );
 			add_action( 'updated_option', array( $this, 'update_last_save_post' ), 10, 3 );
+			add_action( 'generate_rewrite_rules', array( $this, 'generate_rewrite_rules' ) );
```
Comment (bordoni) [inline src/Tribe/Cache_Listener.php]: `[]`

File: src/Tribe/Rewrite.php [hunk 3]
```diff
@@ -637,4 +658,223 @@ protected function get_dynamic_matchers( array $query_vars ) {
...
+	 * @since  TBD
+	 *
+	 * @param string $url              The URLto parse.
+	 * @param array  $extra_query_vars An associative array of extra query vars to use for the parsing. These vars will
+	 *                                 be read before the WordPress defined ones overriding them.
+	 * @param bool   $force Whether to try and use the cache or force a new canonical URL conversion.
+	 *
+	 * @return array An array of query vars, as parsed from the input URL.
+	 */
+	public function parse_request( string $url, array $extra_query_vars = [], $force = false) {
```
Comment (bordoni) [inline src/Tribe/Rewrite.php]: Spacing before `)`

File: src/Tribe/Rewrite.php [hunk 4]
```diff
@@ -637,4 +658,223 @@ protected function get_dynamic_matchers( array $query_vars ) {
...
+
+		// Cache the result for this HTTP request.
+		$this->parse_request_cache[ $url ] = $query_vars;
+
+		return $query_vars;
+	}
+
+	/**
+	 * Dumps the cache before destruction.
+	 */
```
Comment (bordoni) [inline src/Tribe/Rewrite.php]: `@since`

File: src/Tribe/Rewrite.php [hunk 5]
```diff
@@ -637,4 +658,223 @@ protected function get_dynamic_matchers( array $query_vars ) {
...
+					$matched_rule = $match;
+					break;
+				}
+			}
+
+			if ( false !== $matched_rule ) {
+				// Trim the query of everything up to the '?'.
+				$query = preg_replace( '!^.+\?!', '', $query );
+				// Substitute the substring matches into the query.
+				$query               = addslashes( WP_MatchesMapRegex::apply( $query, $matches ) );
```
Comment (bordoni) [inline src/Tribe/Rewrite.php]: Remove indentation.

**Comment by lucatume** [comment]
I've added more code here to make sure we're:
* checking that tags and categories exist before resolving canonical URLs
* cache resolved URls and vars on matches only

File: src/Tribe/Rewrite.php [hunk 6]
```diff
@@ -637,4 +686,231 @@ protected function get_dynamic_matchers( array $query_vars ) {
...
+		$query_vars           = [];
+		$post_type_query_vars = [];
+		$perma_query_vars = [];
+		$url_path             = parse_url( $url, PHP_URL_PATH );
+		if ( empty( $url_path ) ) {
+			$url_path = '/';
+		}
+		// Look for matches, removing leading `/` char.
+		$request_match = ltrim( $url_path, '/' );
+		$url_query = parse_url( $url, PHP_URL_QUERY );
```
Comment (borkweb) [inline src/Tribe/Rewrite.php]: I'd suggest switching to executing `parse_url()` once for efficiency:

```php
$parsed_url = parse_url( $url );
$url_path   = $parsed_url['path'];
$url_query  = $parsed_url['query'];
```

_Perhaps this was copypasta from WP::parse_request()?_

File: src/Tribe/Rewrite.php [hunk 7]
```diff
@@ -637,4 +686,231 @@ protected function get_dynamic_matchers( array $query_vars ) {
...
+		// Fetch the rewrite rules.
+		$rewrite_rules = $this->rewrite->wp_rewrite_rules();
+		$matched_rule = false;
+
+		if ( ! empty( $rewrite_rules ) ) {
+			foreach ( (array) $rewrite_rules as $match => $query ) {
+				$matches_regex = preg_match( "#^$match#", $request_match, $matches )
+				                 || preg_match( "#^$match#", urldecode( $request_match ), $matches );
+
+				if ( $matches_regex ) {
```
Comment (borkweb) [inline src/Tribe/Rewrite.php]: Perhaps reduce a level of nesting by:

```php
if ( ! $matches_regex ) {
  continue;
}
```

_Perhaps this was copypasta from WP::parse_request()?_

File: src/Tribe/Rewrite.php (hunk 6 above)
Comment (borkweb) [reply src/Tribe/Rewrite.php]: This one should probably change regardless of what WP::parse_request is doing.

**Pull request merged**